1) Remove unnecessary object modification in Context.finalize(). If c…

diegomontoya opened the pull request:
…ontext.finalize() is called by GC, the context is already out of scope. Existing code does modification to HashMap and ArrayLists which are only held by context and are going to be GCed anyways. 2) Realm calls Jni-GC aka executeDelayedDisposal on every RealmObject creation so this is a very hot spot. Optimize this by A) using optimized java range loop B) avoid a method call to ArrayList.Clear() 3) Declare a few vars as ArrayList. Doesn't hurt and can only help with performance even if by 0.000001 percent.

bmunkholm commented:
Thanks a lot for the contribution!
Did you validate any performance improvement with measurements?

In general, if the improvement is only 0.000001 % it is definitely not worth it. It actually doesn't come for free. The cost of improvements are the time we spend on reviewing/verifying/testing, which in then not spent on other features that adds more value to users. The other cost is risk of introducing bugs - which is why we spend the time to test :-)

That's just a general comment to any change and in particular performance improvements. It would be lovely if you could test the effect on performance, although it can be very very tricky to do right in Java!

diegomontoya commented:
This comment if for Part 2)  of my commit where every single RealmRow/Object allocation causes a manual JNI-GC into executeDelayedDisposal. Java method calls are not free so reducing the number of method call (when ArrayList.size() == 0, don't call ArrayList.Clear(), -1 method call) is worth the when it is called so many times. 

Benchmark results: (For cases when the 3 array lists are empty)

New is my commit, old is the original executeDelayedDisposal().

08-20 22:10:18.148    3339-3360/? D/executeDelayedNew﹕ 1625685
08-20 22:10:18.153    3339-3360/? D/executeDelayedOld﹕ 2398882
08-20 22:10:19.703    3392-3410/? D/executeDelayedNew﹕ 1704171
08-20 22:10:19.709    3392-3410/? D/executeDelayedOld﹕ 2412596
08-20 22:10:49.182    3450-3468/io.realm.test D/executeDelayedNew﹕ 1842271
08-20 22:10:49.198    3450-3468/io.realm.test D/executeDelayedOld﹕ 7074363
08-20 22:11:02.373    3500-3529/io.realm.test D/executeDelayedNew﹕ 1631586
08-20 22:11:02.379    3500-3529/io.realm.test D/executeDelayedOld﹕ 2481707
08-20 22:11:07.711    3603-3642/? D/executeDelayedNew﹕ 4448522
08-20 22:11:07.727    3603-3642/? D/executeDelayedOld﹕ 7326694
08-20 22:11:13.234    3699-3754/io.realm.test D/executeDelayedNew﹕ 1631959
08-20 22:11:13.240    3699-3754/io.realm.test D/executeDelayedOld﹕ 2369423
08-20 22:11:18.000    3801-3824/? D/executeDelayedNew﹕ 1746719
08-20 22:11:18.008    3801-3824/? D/executeDelayedOld﹕ 2307235
08-20 22:11:29.509    3863-3883/io.realm.test D/executeDelayedNew﹕ 4151791
08-20 22:11:29.524    3863-3883/io.realm.test D/executeDelayedOld﹕ 5272860
08-20 22:11:37.888    3914-3932/io.realm.test D/executeDelayedNew﹕ 5203016
08-20 22:11:37.900    3914-3932/io.realm.test D/executeDelayedOld﹕ 7052860
08-20 22:11:46.169    3966-3987/? D/executeDelayedNew﹕ 4926731
08-20 22:11:46.186    3966-3987/? D/executeDelayedOld﹕ 9636767
08-20 22:11:54.746    4029-4047/io.realm.test D/executeDelayedNew﹕ 1941791
08-20 22:11:54.755    4029-4047/io.realm.test D/executeDelayedOld﹕ 2766250

For the following I changed some var scopes to access the methods/vars directly to better reveal actual bench values instead of going through another proxy method.

Added code to below setup in JNITableTest.java.

    public void testBenchMethod1() {
        long start = System.nanoTime();

        for (int i = 0; i < 10000; i++) {
            t.context.executeDelayedDisposal();
        }

        long timeSpent = System.nanoTime() - start;

        Log.d("executeDelayedNew", String.valueOf(timeSpent));
    }

    public void testBenchMethod2() {
        long start = System.nanoTime();

        for (int i = 0; i < 10000; i++) {
            t.context.executeDelayedDisposalOld();
        }

        long timeSpent = System.nanoTime() - start;

        Log.d("executeDelayedOld", String.valueOf(timeSpent));
    }

diegomontoya commented:
@bmunkholm My 0.000001 comment was for part 3) of the commit which on second though does really nothing except make me feel better. =)

You guys can do some benchmarks regarding the part 1) which introduce a special JNI-GC path just for Finalize().

diegomontoya commented:
Btw. I see a failed test, though it passes all the tests on my end. What is the failed test?

cmelchior commented:
@diegomontoya Our CI server had some connection problems to our test device. It doesn't seem to a be a problem in your PR

cmelchior commented on realm/src/main/java/io/realm/internal/Context.java:
```diff
@@ -44,25 +44,50 @@
-            for (int i = 0; i < abandonedTables.size(); i++) {
-                long nativePointer = abandonedTables.get(i);
+            if(abandonedTables.size() > 0) {
+                for (long nativePointer : abandonedTables) {
```
You should not use a `for...in` loop here as it will allocate a Iterator under the hood. It is safer to use:

```
int tableViewSize = abandonedTableViews.size()
 for (int i = 0; i < tableViewSize ; i++) {
  // ...
}
```

cmelchior commented on realm/src/main/java/io/realm/internal/Context.java:
```diff
@@ -44,25 +44,50 @@
         synchronized (this) {
-            for (int i = 0; i < abandonedTables.size(); i++) {
-                long nativePointer = abandonedTables.get(i);
+            if(abandonedTables.size() > 0) {
```
Small style issue: Space between `if (`

cmelchior commented on realm/src/main/java/io/realm/internal/Context.java:
```diff
@@ -44,25 +44,50 @@
+                abandonedTables.clear();
+            }
+
+            if(abandonedTableViews.size() > 0) {
```
`if (`

cmelchior commented on realm/src/main/java/io/realm/internal/Context.java:
```diff
@@ -44,25 +44,50 @@
+            }
+
+            if(abandonedTableViews.size() > 0) {
+                for (long nativePointer : abandonedTableViews) {
```
Same...

cmelchior commented on realm/src/main/java/io/realm/internal/Context.java:
```diff
@@ -44,25 +44,50 @@
+                abandonedTableViews.clear();
+            }
+
+            if(abandonedQueries.size() > 0) {
```
Same...

cmelchior commented on realm/src/main/java/io/realm/internal/Context.java:
```diff
@@ -44,25 +44,50 @@
+            }
+
+            if(abandonedQueries.size() > 0) {
+                for (long nativePointer : abandonedQueries) {
```
Same...

kneth commented on realm-jni/build.gradle:
```diff
@@ -147,7 +147,7 @@ targets.each { target ->
         environment PATH: "${buildDir}/standalone-toolchains/${target.toolchain.name}/bin:${System.env.PATH}"
         environment CC: "${target.toolchain.commandPrefix}-${clang?'clang':'gcc'}"
-        environment STRIP: "${target.toolchain.commandPrefix}-strip -o librealm-jni-${target.name}-stripped.so"
+        environment STRIP: "${target.toolchain.commandPrefix}-strip -o libearthworm-${target.name}-stripped.so"
```
Why do we have to change the name of the file?

kneth commented on realm/build.gradle:
```diff
@@ -26,6 +29,8 @@ android {
 
 dependencies {
+    compile 'com.android.support:multidex:1.0.0'
+    compile 'it.unimi.dsi:fastutil:7.0.7'
```
Is adding a dependency to an external library a good idea? @cmelchior @nhachicha do you have experiences with it?

diegomontoya replied:
Ooops. This was just my local test changes. Never meant them to show up here.

diegomontoya commented:
@kneth Ignore thhose commits. Didn't know the changes after my pull commit is actually going to show up here. Assumed the pull request is "just this commit". The following commits in my fork are actually just my local test/changes.  I might have to fork another just to keep it clean. I will close this pull now since this is commit tree is pretty polluted.

kneth commented:
@diegomontoya OK. At least I got to study `fastutil`which looks like an interesting library :-)
